RobustTrees/RobustOCT.py

```diff
@@ -0,0 +1,195 @@
+            self.y = y.values
+            self.labels = np.unique(self.y)
+        else:
+            self.y = y
+            self.labels = np.unique(self.y)
+        # datapoints contains the indicies of our training data
+        self.datapoints = np.arange(0, self.X.shape[0])
+        self.tree = tree
+
+        '''
+        Get range of data, and store indices of branching variables based on range
+        '''
+        min_values = X.min(axis=0)
+        max_values = X.max(axis=0)
+        f_theta_indices = []
+        b_indices = []
+        for f in self.cat_features:
+            min_value = min_values[f]
+            max_value = max_values[f]
+
+            # cutoffs are from min_value to max_value - 1
+            for theta in range(min_value, max_value): 
+                f_theta_indices += [(f,theta)]
+                b_indices += [(n, f, theta) for n in self.tree.Nodes]
+
+        self.min_values = min_values
+        self.max_values = max_values
+        self.f_theta_indices = f_theta_indices
+        self.b_indices = b_indices
+
+        '''
+        Create uncertainty set
+        NOTE - I think that we should have users somehow input the uncertainty
+        parameters rather than us do it. Perhaps we can add functionality to
+        generate uncertain parameters through the below process
+        '''
+        if seed:
+            np.random.seed(seed)
+
+        # Create epsilon
+        if l > 0:
+            self.epsilon = -1 * math.log(l) * len(self.datapoints)
+        else: # l == 0
+            self.epsilon = sys.maxsize # Infinity
+
+        # Get q_f values for each feature
+        q_f = np.random.normal(loc=q, scale=s, size=len(self.cat_features))
+        # Snap q_f to range [0,1]
+        q_f[q_f <= 0] = np.nextafter(np.float32(0), np.float32(1))
+        q_f[q_f > 1] = 1.0
+        self.q_values = q_f
+
+        # Create gammas (budget of feature uncertainty)
+        gammas_df = copy.deepcopy(X)
+        gammas_df = gammas_df.astype('float')
+        for f in range(len(q_f)):
+            if q_f[f] == 1:
+                gammas_df[gammas_df.columns[f]] = self.epsilon + 1
+            else:
+                gammas_df[gammas_df.columns[f]] = -1 * math.log(1 - q_f[f])
+
+        self.gammas = gammas_df
+
+        # Create etas (budget of label uncertainty)
+        if p == 1:
+            self.eta = self.epsilon + 1
+        else:
+            self.eta = math.log(p / (1 - p))
+
+        # Decision Variables
+        self.t = 0
+        self.b = 0
+        self.w = 0
+
+        # Gurobi model
+        self.model = Model('RobustOCT')
+        # The cuts we add in the callback function would be treated as lazy constraints
+        self.model.params.LazyConstraints = 1
+
+        '''
+        To compare all approaches in a fair setting we limit the solver to use only one thread to merely evaluate 
+        the strength of the formulation.
+        '''
+        # self.model.params.Threads = 1
```

Comment:
Yes. From the [Gurobi reference manual](https://www.gurobi.com/wp-content/plugins/hd_documentations/documentation/9.0/refman.pdf):
> It will generally use all of the cores in the machine, but it may choose to use fewer.

 We limited it to 1 thread to have consistency of runtimes between instances for comparison, but I assume that we would want users to be able to set this themselves or use the default.